### PR TITLE
bpo-42979: Enhance abstract.c assertions checking slot result

### PR DESCRIPTION
**Pull request by vstinner**
Add _Py_CheckSlotResult() function which fails with a fatal error if
a slot function succeeded with an exception set or failed with no
exception set: write the slot name, the type name and the current
exception (if an exception is set).

https://bugs.python.org/issue42979

### CODE REVIEW

**Comment by vstinner** [comment]
Example which tries to mimick bpo-42979 by injecting `PyErr_SetString(PyExc_ValueError, "bug")` before `PyType_Ready()` is called in zoneinfo exec function:

```
Fatal Python error: _Py_CheckSlotResult: Slot __getitem__ of type dict succeeded with an exception set
Python runtime state: initialized
ValueError: bug

Current thread 0x00007ff43e1b1740 (most recent call first):
  File "<frozen importlib._bootstrap>", line 241 in _call_with_frames_removed
  File "<frozen importlib._bootstrap_external>", line 1134 in exec_module
  File "<frozen importlib._bootstrap>", line 698 in _load_unlocked
  File "<frozen importlib._bootstrap>", line 729 in _load
  File "/home/vstinner/python/master/./setup.py", line 620 in check_extension_import
  File "/home/vstinner/python/master/./setup.py", line 475 in build_extensions
  File "/home/vstinner/python/master/Lib/distutils/command/build_ext.py", line 340 in run
  File "/home/vstinner/python/master/Lib/distutils/dist.py", line 985 in run_command
  File "/home/vstinner/python/master/Lib/distutils/cmd.py", line 313 in run_command
  File "/home/vstinner/python/master/Lib/distutils/command/build.py", line 135 in run
  File "/home/vstinner/python/master/Lib/distutils/dist.py", line 985 in run_command
  File "/home/vstinner/python/master/Lib/distutils/dist.py", line 966 in run_commands
  File "/home/vstinner/python/master/Lib/distutils/core.py", line 148 in setup
  File "/home/vstinner/python/master/./setup.py", line 2632 in main
  File "/home/vstinner/python/master/./setup.py", line 2662 in <module>
```

File: Include/cpython/abstract.h [hunk 1]
```diff
@@ -39,6 +39,12 @@ PyAPI_FUNC(PyObject *) _Py_CheckFunctionResult(
     PyObject *result,
     const char *where);
 
+// Usage: assert(_Py_CheckSlotResult(obj, "__getitem__", result != NULL)));
+PyAPI_FUNC(int) _Py_CheckSlotResult(
```
Comment (markshannon) [inline Include/cpython/abstract.h]: Why an API function?
You know that you will regret it later :wink: https://www.python.org/dev/peps/pep-0620/

**Comment by markshannon** [comment]
LGTM, apart from the API creep.

File: Include/cpython/abstract.h (hunk 1 above)
Comment (vstinner) [reply Include/cpython/abstract.h]: Oh, let me see that PEP... ah yes ;-) I wanted to define it close to _Py_CheckFunctionResult() definition, but you're right that it's a bad idea to export the symbol.

**Comment by vstinner** [comment]
> the API creep

Fixed: I moved the function definition to the internal C API. I also added an unit test.

Note: this change has no impact on a release build (when Python is built with NDEBUG macro defined).

**Pull request merged**